### PR TITLE
Audit Sha1() usage

### PR DESCRIPTION
**Pull request by lesleyjanenorton**
Replaces our single `sha1()` function with `getSha1ForHIBP()` and `getLowerCasedSha1()` which should return hashes correctly formatted for the purpose. This will hopefully help us cut down on the .toUpperCase()/.toLowerCase()/.toBasketCase() shenanigans. 

`getSha1ForHIBP(emailAddress)` 
- converts emailAddress to lowercase letters before hashing
- converts the hash to uppercase letters.
- should be used anytime we're creating a hash for scanning HIBP

`getLowerCaseSha1(emailAddress)` 
- converts emailAddress to lowercase letters before hashing
- leaves the hash in lowercase letters to be stored in our DB.
- should be used anytime a hash is being created for storing in our DB, or for finding something in our DB. 

TODO:
It would be great to add some related tests and a readme documenting the Laws of Monitor Casing.

### CODE REVIEW

File: sha1-utils.js [hunk 1]
```diff
@@ -2,8 +2,23 @@
 
 const crypto = require("crypto");
 
-function getSha1(email) {
-  return crypto.createHash("sha1").update(email).digest("hex");
+// Email addresses must be lowercased before hashing
+// Hashes sent to HIBP should be uppercased
+function getSha1ForHIBP (email) {
+  const formattedEmail = email.toLowerCase();
+  const emailHash = crypto.createHash("sha1").update(formattedEmail).digest("hex");
```
Comment (pdehaan) [inline sha1-utils.js]: Silly Q: Thoughts on replacing some of this w/ `return getLowerCaseSha1(email).toUpperCase()`? and keeping it somewhat DRYer?

File: scripts/subscribe-lowercase-hashes.js [hunk 2]
```diff
@@ -11,8 +11,7 @@ const getSha1 = require("../sha1-utils");
 
 
 async function subscribeLowercaseHashToHIBP(emailAddress) {
-  const lowerCasedEmail = emailAddress.toLowerCase();
-  const lowerCasedSha1 = getSha1(lowerCasedEmail);
+  const lowerCasedSha1 = getSha1.getSha1ForHIBP(emailAddress);
```
Comment (pdehaan) [inline scripts/subscribe-lowercase-hashes.js]: Q: the variable name is `lowerCasedSha1`, but I believe that `getSha1ForHIBP()` returns an uppercased string.

**Comment by pdehaan** [comment]
Another interesting fact, it doesn't [currently] look like any files use both the `.getSha1ForHIBP()` _and_ `.getLowerCaseSha1()` APIs.

We could potentially do something like:

```js
const sha1 = require("../sha1-utils").getSha1ForHIBP;
```

Which just pulls the single method we need and aliases it to `sha1()`.

**PROS:** The PR diff would be minimal.
**CONS:** Makes it a bit less obvious what the code is doing I suppose, and would make it a bit more awkward if in some places we did eventually need to use both the `getSha1ForHIBP()` and `getLowerCaseSha1()` methods in the same file.

I'm just thinking out loud here, don't feel like this is a good, or even mediocre solution.

@groovecoder Have any other ideas or suggestions here?

**Comment by groovecoder** [comment]
This is a great and thorough audit! I had a bunch of line-by-line comments, but I think a summary here is better ...

@pdehaan - I wouldn't want to use `const sha1 = require("../sha1-utils").[getSha1ForHIBP|getLowerCaseSha1]` - it seems like another new point where confusion or accidents could creep into the code.

This PR changes all of the _calling code_ to use explicit `getLowerCasedSha1()` and `getSha1ForHIBP()` functions, but there's probably a more robust way to enforce the 2 major rules we have:

* For HIBP, upper-case the hash before sending API calls
* For database, lower-case the hash before writing or reading

It's probably better to also enforce the proper casing of hashes at the HIBP & DB touch-points.

E.g., the `hibp` module has just 2 functions that actually make use of the hash: `getBreachesForEmail` and `subscribeHash`. If we enforce `toUpperCase` in those functions, we can be sure the hash will be upper-cased before it goes to HIBP, even if we forget to call `getSha1ForHIBP` in the calling code first. (If we send an upper-case hash, there's no harm in calling `toUpperCase` on it again.)

Similarly, the `DB` module has 9 places that actually reference the `subscribers.primary_sha1` or the `email_addresses.sha1` fields. If we enforce `toLowerCase` in those functions, we can be sure the hash will be lower-cased before it goes into the DB, even if we forget to call `getLowerCasedSha1` in the calling code first. (If we send a lower-case hash, there's no harm in calling `toLowerCase` on it again.)

**Comment by lesleyjanenorton** [comment]
Would it be too much of a nightmare to uppercase the hashes in our db and then simply add `.toUpperCase()` in our existing sha1 function?

File: scripts/subscribe-lowercase-hashes.js (hunk 2 above)
Comment (lesleyjanenorton) [reply scripts/subscribe-lowercase-hashes.js]: It's totally confusing. I think the `lowerCased` bit refers to the letters of the email address being lowercased before being encrypted.

**Comment by groovecoder** [comment]
It wouldn't be technically hard. I think it would simply be:

```sql
UPDATE subscribers set primary_sha1 = upper(primary_sha1);
UPDATE email_addresses set sha1 = upper(sha1);
```

But it's a bit risky to try to update all the data in the entire DB. 😐

**Comment by groovecoder** [comment]
Not planning to merge this.

**Pull request closed without merging**